### PR TITLE
glfw: 3.3.4 -> 3.3.5 and fix linkage with X11

### LINKED ISSUE #142583: GLFW_CURSOR_DISABLED does not hide cursor on X11
### Describe the bug
`glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_DISABLED)` should hide the cursor but does not on X11.  I have not tested this on Wayland or Cocoa.

### Steps To Reproduce
1. Fetch GLFW 3.3.x (to match unstable, which is currently 3.3.4)
2. With GLFW's dependencies (`cmake xorg.libX11 xorg.libXrandr libGL xorg.libXinerama xorg.libXcursor xorg.libXi xorg.libXext` on X11), build the library's examples.  The default CMake build should do the job:

   ```bash
   mkdir build
   cd build
   cmake ..
   cmake --build . -j `nproc`
   cd - # reset cwd
   ```
3. Execute `build/examples/wave` using `steam-run` or equivalent.  Verify that left clicking the window hides the cursor since `GLFW_CURSOR_DISABLED` is set.
4. With NixOS's `glfw`, build the example manually with `gcc -Ideps examples/wave.c -lglfw -lGL`.  Execute it and now clicking moves the cursor to the center of the window but does not hide it.

### Expected behavior
NixOS build should behave the same as the hand-built one.

### Additional context
GLFW_CURSOR_HIDDEN is also affected.

### Notify maintainers
@MarcWeber and @Twey

### Metadata
```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.69, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11pre322215.9bf75dd50b7"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by McSinyx**
###### Motivation for this change
Fix GH-142583 (thanks, @symphorien!), plus supersede and thus close GH-143709

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/glfw/3.x.nix
Comment (symphorien) [inline pkgs/development/libraries/glfw/3.x.nix:19]: please add a comment about the necessity of the patch.
Can a user need both patches at the same time ? (for example when targetting XWayland ? I'm not familiar with glfw, it might not make sense.)

Note that would only minimally increase closure size as wayland support adds a dependency on libxkbcommon which brings libx11 and libxcb in already.

Comment (McSinyx) [reply pkgs/development/libraries/glfw/3.x.nix:21]: > please add a comment about the necessity of the patch.

Done!

> Can a user need both patches at the same time? (for example when targetting XWayland?)

IIUC the Wayland patch is apply for a file that's only compiled when Wayland support is specified and the X11 patch only affects linking instructions for building for X11, and XWayland is for running the X11 build on Wayland, so the two patches should be mutually exclusive.

**Review by symphorien — APPROVED** [review]
let's wait for maintainer feedback a bit longer.

**Pull request merged**
